### PR TITLE
Added recently used applications widget

### PR DESCRIPTION
**Pull request by abujeda**
Recently used applications widget based on the list of application data cache files

Fixes: https://github.com/OSC/ondemand/issues/2477

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1135148780858012/1203796685604202) by [Unito](https://www.unito.io)

### CODE REVIEW

**Comment by adaybujeda** [comment]
Which header do you prefer @johrstrom?

Screenshot 2023-01-22 at 17 01 06

Screenshot 2023-01-22 at 17 00 50

**Comment by johrstrom** [comment]
Probably the latter because it's in line with the other widgets.

I was off on Friday, I'll take a look at this today.

**Comment by abujeda** [comment]
OK - will update to have that header

File: apps/dashboard/app/helpers/recently_used_apps_helper.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+# Helper methods for RecentlyUsedApps widget
+module RecentlyUsedAppsHelper
+  # Returns the 4 most recently used interactive applications based on their cache file
+  # @return [Array<BatchConnect::App>] The 4 most recently used apps
+  def recently_used_apps
+    cache_files = BatchConnect::Session.cache_root.children.sort_by do |pathname|
+                    File.mtime(pathname)
+                  end.reverse.map { |pathname| pathname.basename.to_s }.slice(0, 4)
```
Comment (johrstrom) [inline apps/dashboard/app/helpers/recently_used_apps_helper.rb:10]: Let's try to keep uniformity here in these closures.

```suggestion
                  end.reverse.map do |pathname| 
                    pathname.basename.to_s 
                   end.slice(0, 4)
```

File: apps/dashboard/app/helpers/recently_used_apps_helper.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+# Helper methods for RecentlyUsedApps widget
+module RecentlyUsedAppsHelper
+  # Returns the 4 most recently used interactive applications based on their cache file
+  # @return [Array<BatchConnect::App>] The 4 most recently used apps
+  def recently_used_apps
+    cache_files = BatchConnect::Session.cache_root.children.sort_by do |pathname|
```
Comment (johrstrom) [inline apps/dashboard/app/helpers/recently_used_apps_helper.rb:8]: This looks expensive to compute. Maybe we should put this within a cache block.

```ruby
Rails.cache.fetch('recently_used_apps', expires_in: 1.hour) do

end
```

There could be more effective caching schemes, i.e., invalidate it when one of the cache files is updated, but I feel like we can defer that.

Comment (abujeda) [reply apps/dashboard/app/helpers/recently_used_apps_helper.rb:17]: Added caching.

File: apps/dashboard/app/helpers/recently_used_apps_helper.rb (hunk 1 above)
Comment (abujeda) [reply apps/dashboard/app/helpers/recently_used_apps_helper.rb]: fixed

File: apps/dashboard/app/helpers/recently_used_apps_helper.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  # Returns the 4 most recently used interactive applications based on their cache file
+  # @return [Array<BatchConnect::App>] The 4 most recently used apps
+  def load_recently_used_apps
+    cache_files = BatchConnect::Session.cache_root.children.sort_by do |pathname|
+                    File.mtime(pathname)
+                  end.reverse.map do |pathname|
+                    pathname.basename.to_s
+                  end.slice(0, 4)
+    sys_apps_index = {}
+    SysRouter.apps.select(&:batch_connect_app?).each do |ood_app|
```
Comment (johrstrom) [inline apps/dashboard/app/helpers/recently_used_apps_helper.rb:23]:
```suggestion
   (SysRouter.apps + DevRouter.apps + UsrRouter.apps).select(&:batch_connect_app?).each do |ood_app|
```

We may as well add this here and now. I'm not super concerned about `dev` apps, but there are sites that use `usr` apps almost exclusively. That is, they don't really even have `sys` apps, they just share all their apps from the staff and/or regular users.

Comment (abujeda) [reply apps/dashboard/app/helpers/recently_used_apps_helper.rb]: Done. Updated to use `@sys_apps + @dev_apps + @usr_apps` that are initialized in the `ApplicationController` class.

Comment (johrstrom) [reply apps/dashboard/app/helpers/recently_used_apps_helper.rb]: That'll work for now.  IIRC member variables are generated every page load so caching the member variables is a little funny, but that's OK (let's hope so anyhow).

**Review by johrstrom — APPROVED** [review]
This works for me. You can squash if you like.

**Comment by abujeda** [comment]
Great! Squash completed. Ready to go 🚀 .

Will start working on the documentation updates.

**Review by johrstrom — APPROVED** [review]

**Pull request merged**